### PR TITLE
add @namespace annotation support

### PR DESCRIPTION
**Pull request by FredKSchott**
- [X] CHANGELOG.md has been updated
- Creates a new 'namespace' feature to represent namespaces
- Detects namespaces tagged with `@namespace`
- Warns if multiple namespaces are defined
- Next up: adding `@memberof` support for connecting members to namespaces

/cc @justinfagnani @rictic @usergenic

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,6 +12,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 -->
 
 <!--## Unreleased-->
+* Add `@namespace` annotation support.
```
Comment (rictic) [inline CHANGELOG.md]: Consider mentioning that it's a jsdoc tag. Maybe:

```
* Support the `@namespace` jsdoc tag.
```

File: src/polymer/namespace-scanner.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+   * Look for object assignments with @namespace in the docs.
+   */
+  enterAssignmentExpression(
+      node: estree.AssignmentExpression, parent: estree.Node) {
+    this._initNamespace(parent, () => esutil.objectKeyToString(node.left)!);
+  }
+
+  private _initNamespace(node: estree.Node, getName: () => string) {
+    const comment = esutil.getAttachedComment(node) || '';
+    // Quickly filter down to potential candidates.
```
Comment (rictic) [inline src/polymer/namespace-scanner.ts]: +1

File: src/polymer/namespace-scanner.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    if (!comment || comment.indexOf('@namespace') === -1) {
+      return;
+    }
+
+    const docs = jsdoc.parseJsdoc(comment);
+    const name = jsdoc.getTag(docs, 'namespace', 'name') || getName();
+    if (!name) {
+      throw new Error(`Unable to determine name for @namespace: ${comment}`);
```
Comment (rictic) [inline src/polymer/namespace-scanner.ts]: We don't want to throw errors in the analyzer based on unexpected source, both because it's a bad user experience, and because the code doing the analysis may not care and want the analysis result anyways.

Instead, we should generate a Warning and propagate it up. We haven't needed it yet, but I've almost written a WarningFeature (a Feature that only carries a Warning/array of Warnings) a couple of times for scanners to emit when they can't emit their usual scanned feature type but they still found a problem.

File: src/polymer/namespace-scanner.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    const sourceRange = this.document.sourceRangeForNode(node);
+    if (!sourceRange) {
+      throw new Error(
+          `Unable to determine sourceRange for @namespace: ${comment}`);
```
Comment (rictic) [inline src/polymer/namespace-scanner.ts]: This is fine though, because this really should never happen (and we haven't yet found a case when it doesn't happen).

File: src/polymer/namespace-scanner.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+   * Look for object declarations with @namespace in the docs.
+   */
+  enterVariableDeclaration(
+      node: estree.VariableDeclaration, _parent: estree.Node) {
+    if (node.declarations.length !== 1) {
+      return;  // Ambiguous.
+    }
+    this._initNamespace(node, () => {
+      const id = node.declarations[0].id;
+      return esutil.objectKeyToString(id)!;
```
Comment (rictic) [inline src/polymer/namespace-scanner.ts]: They're very similar, but I'd recommend ast-value.getIdentifierName here instead of objectKeyToString, only because the former was designed for this purpose, and we may extend objectKeyToString in the future in ways unexpected to this code.

File: src/polymer/namespace.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  name: string;
+  kinds: Set<string>;
+  identifiers: Set<string>;
+  sourceRange: SourceRange;
+  astNode: any;
+  warnings: Warning[];
+
+  constructor(scannedNamespace: ScannedNamespace) {
+    this.name = scannedNamespace.name;
+    this.kinds = new Set(['namespace']);
```
Comment (rictic) [inline src/polymer/namespace.ts]: It's nice that this maps so well onto the kinds and identifiers system.

Please add the `'namespace': Namespace` mapping to the kind map in model/document.ts

File: src/test/analyzer_test.ts [hunk 7]
```diff
@@ -532,6 +532,45 @@ suite('Analyzer', () => {
...
+    ]);
+    assert.deepEqual(
+        document.getWarnings({imported: true}), [{
+          message:
+              'Found more than one namespace named ExplicitlyNamedNamespace.',
+          severity: Severity.WARNING,
+          code: 'multiple-polymer-namespaces',
+          sourceRange: {
+            file: 'static/namespaces/namespace-duplicate.js',
+            start: {column: 0, line: 5},
```
Comment (rictic) [inline src/test/analyzer_test.ts]: Here and elsewhere in this PR, asserting on source ranges is really brittle. If we use CodeUnderliner the code is easier to read, easier to maintain, and isn't broken by adding unrelated lines to a fixture file.

You can use assertSubset to leave out the sourceRange property of the expected object.

File: src/polymer/namespace.ts (hunk 6 above)
Comment (FredKSchott) [inline src/polymer/namespace.ts]: Yea I was really impressed as well.

Already added it to `FeatureKinds`, let me know if you meant somewhere else that I've missed.

File: src/test/analyzer_test.ts (hunk 7 above)
Comment (FredKSchott) [inline src/test/analyzer_test.ts]: yea just corrected that in your last PR, will do the same here

File: src/polymer/namespace-scanner.ts (hunk 3 above)
Comment (FredKSchott) [inline src/polymer/namespace-scanner.ts]: I actually couldn't think of how name could ever be falsey here. Instead of throwing, maybe just ignoring and returning nothing is a better solution until we have a way to propagate real Warnings?

A lot of this is based on behavior in Behavior scanning, so throwing doesn't introduce a new problem.

File: src/polymer/namespace-scanner.ts (hunk 5 above)
Comment (FredKSchott) [inline src/polymer/namespace-scanner.ts]: +1 I'll update in behavior scanner as well

Comment (FredKSchott) [inline src/polymer/namespace-scanner.ts]: ~~+1 I'll update in behavior scanner as well~~ Actually scratch that, it looks like `getIdentifierName` doesn't support everything that `objectKeyToString` does. I'll leave behavior scanner as is for now.

Comment (justinfagnani) [inline src/polymer/namespace-scanner.ts]: Either will be fine for now since this is a variable declaration, but both do some possibly surprising things and don't handle symbols, and we need to be careful in other cases.

File: src/polymer/namespace-scanner.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  /**
+   * Look for object assignments with @namespace in the docs.
+   */
+  enterAssignmentExpression(
+      node: estree.AssignmentExpression, parent: estree.Node) {
+    this._initNamespace(parent, () => esutil.objectKeyToString(node.left)!);
+  }
+
+  private _initNamespace(node: estree.Node, getName: () => string) {
+    const comment = esutil.getAttachedComment(node) || '';
```
Comment (justinfagnani) [inline src/polymer/namespace-scanner.ts]: I don't think you need the fallback

File: src/polymer/namespace-scanner.ts (hunk 3 above)
Comment (justinfagnani) [inline src/polymer/namespace-scanner.ts]: It introduces a new place we need to clean up to keep the analyzer from dying on normal operation. We need to only throw on unrecoverable internal errors, there's even a bug on the behavior exception.

File: src/polymer/namespace.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+ */
+
+import * as estree from 'estree';
+
+import * as jsdoc from '../javascript/jsdoc';
+import {Document, Feature, Resolvable, SourceRange} from '../model/model';
+import {Severity, Warning} from '../warning/warning';
+
+/**
+ * The metadata for a Polymer namespace.
```
Comment (justinfagnani) [inline src/polymer/namespace.ts]: Polymer -> JavaScript

File: src/polymer/namespace.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  resolve(document: Document) {
+    const foundNamespaces = document.getById(
+        'namespace', this.name, {imported: true, externalPackages: true});
+    if (foundNamespaces.size > 0) {
+      document.warnings.push({
+        message: `Found more than one namespace named ${this.name}.`,
+        severity: Severity.WARNING,
+        code: 'multiple-polymer-namespaces',
```
Comment (justinfagnani) [inline src/polymer/namespace.ts]: polymer -> javascript

File: src/test/static/namespaces/import-all.html [hunk 11]
```diff
@@ -0,0 +1,4 @@
+<script src="namespace-named.js"></script>
```
Comment (justinfagnani) [inline src/test/static/namespaces/import-all.html]: I don't think we need these html files, just test on the JS files.

File: src/polymer/namespace-scanner.ts (hunk 3 above)
Comment (FredKSchott) [inline src/polymer/namespace-scanner.ts]: This is only ever an issue for nonstandard, dynamic namespaces like `DynamicNamespace['ArrayNotation'] `. 

This logic was based on Behavior scanning, so this isn't a new problem. Instead of throwing, maybe just ignoring and returning nothing is a better solution until we have a way to propagate real Warnings? I'll leave it as-is for now.

File: src/polymer/namespace-scanner.ts (hunk 4 above)
Comment (FredKSchott) [inline src/polymer/namespace-scanner.ts]: Yea exactly, `sourceRange` was `SourceRange | undefined` and I wanted to explicitly check rather than sending it into ScannedNamespace with a `!`. Hopefully we'll never hit this, but if we do at least we'll know immediately.

File: src/test/static/namespaces/import-all.html (hunk 11 above)
Comment (FredKSchott) [inline src/test/static/namespaces/import-all.html]: I test the JS files directly during scanning, but for full analysis its necessary to analyze them all together to test for duplicates across files. Will also be useful to test `memberof` across files.

File: src/polymer/namespace-scanner.ts (hunk 8 above)
Comment (FredKSchott) [inline src/polymer/namespace-scanner.ts]: good catch

File: src/test/static/namespaces/import-all.html (hunk 11 above)
Comment (justinfagnani) [inline src/test/static/namespaces/import-all.html]: ok
